Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -1,217 +1,269 @@
-  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
-  
-If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
++ `System.Runtime.Intrinsics.Arm` is used type definitions shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Arm64` is used for type definitions for the `ARM64` platform
+  + The primary implementation of `ARM64` intrinsics wil occur within this namespace
+  + While `x86` and `x64` share a common namespace.  This document is recommending a separate namespace
+  for `ARM32` and `ARM64`.  This is because `AARCH64` is a separate `ISA` from the `AARCH32` `Arm` & `Thumb`
+  instruction sets.  It is not an `ISA` extension, but rather a new `ISA`.  This is different from `x64`
+  which could be viewed as a superset of `x86`.
+  + The logical grouping of `ARM64` and `ARM32` instruction sets is different.  It is controlled by
+  different sets of `System Registers`.
 
-+ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+For the convenience of the end user, it may be useful to add convenience API's which expose functionality
+which is common across platforms and sets of platforms.  These could be implemented in terms of the
+platform specific functionality.  These API's are currently out of scope of this initial design document.
 
-If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+### Logical Set Class Names
 
-+ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+Within the `System.Runtime.Intrinsics.Arm.Arm64` namespace there will be a separate `static class` for each
+logical set of instructions
 
-Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+The sets will be chosen to match the granularity if the `ARM64` `ID_*` register fields.
 
-#### Names
+#### Specific Class Names
 
-Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+TBD Add table to document each known extension and recommended `static class` name.
 
-`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+### Intrinsic Method Names
 
-### Intrinsic Interface Documentation
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named
+assembly instructions.
 
-+ Documentation will be minimal preferring the underlying ARM documentation
-+ Namespaces will briefly document corresponding specification
-+ Intrinsic methods will briefly document corresponding assembly instruction(s) for each platform
+Where precedence exists within the `System.Runtime.Intrinsics.X86` namespace, identical method names will be
```

Comment:
It may be worth noting that, for the `X86` intrinsics. `Packed` instructions have no special suffix. `Scalar` instructions (those that take a `Vector128<T>` or `Vector256<T>` argument, but only operate on the first value in the register) are going to be suffixed with `Scalar`.

Ex: `Vector128<float> Add(Vector128<float> left, Vector128<float> right)` and `Vector128<float> AddScalar(Vector128<float> left, Vector128<float> right)`

Reply:
OK